### ISSUE TITLE
Penalize side of street change

### ISSUE THREAD

**Issue by nilsnolde** [valhalla/valhalla#4443]
I had a request to penalize when origin/destination are on different sides of the street, which is obviously only relevant for singly digitized roads which are not oneways. This can be important for e.g. delivering mail or garbage trucks etc. E.g.

```
     ---                          ---
    | A |                        | C |
     ---                          ---
x-----a------x----b---------x------c-------x
                 ---
                | B |
                 ---       
```

where `A/B/C` are house centroids as input, `a/b/c` how Valhalla snaps the input, `x` are graph nodes.

If a VRP looks like this, the current matrix code would result in A -> B -> C, which is really zig-zagging the final route (imagine many more addresses this quickly becomes really crappy). If we could penalize the two street changes, the optimization would have a chance to do A -> C -> B, which would result in only one street side change.

I'd advocate to just increase the final time if that happens, not even impact the cost (there's no point I think, the only way to avoid a street side change in the algo would generally be a u-turn). WDYT?

**Comment by kevinkreiser** [comment]
i agree that this is a valuable problem to solve. NDA prevents me from disclosing more but ive also considered this in the past for certain use-cases. i have 2 primary comments:
 
 1. regarding the application of such a penality, i dont think i would make the penalty a binary thing. that is, i dont think id always apply it or always not apply it. one needs to consider the locality before deciding whether changing sides matters or not. concretely, you can imagine many scenarios where its more efficient to take the optimized route that we give today over, lets say, driving down the road to stay on one side and driving back to get the other side. pretty much all rural roads are better off just driving the road in one direction and easily just crossing the street (by foot) while parking on the side you are driving on. in other words, penalizing this only makes sense if its likely that you'd have a hard time crossing the street. even in little towns it might be fine to cross the street to do a delivery. so one thing to think about is do we have other metadata in the graph that can clue us in to whether or not we'd actually want to penalize side of street changes. you could think of road class or maybe network density but its a hard problem imho.
 2. how do we actually enact the penalty? impelmentation-wise i mean. maybe you were saying we can do it at the top level optimization step with a time-only penalty, somethign like its costs me 2mins to cross the street and cross back? what other options do we have there?

**Comment by nilsnolde** [comment]
Yeah, I also thought that it might be a bit presumptuous to apply it in every case. On the other hand, as you say, it'll be non-trivial to even come up with a proxy to decide that. IMO it'd be ok to include in a first pass without going down that rabbit hole.

I didn't even think of our optimization, rather downstream processing:) Not 100% sure yet where to put it. I guess in every action that wants/needs it (route/matrix?). Similar to `waiting_secs` for routing with intermediate waypoints maybe, where we add it after the fact before returning? But yeah, just a fixed time-only penalty would suffice IMO.

**Comment by radekvermirovsky** [comment]
1)
It would be remarkable to intelligently determine whether to apply a penalty. However, IMHO, this is an inherently challenging task. In our scenarios, we almost always know whether to impose a penalty based on the nature of optimization and the specific problem. Of course, we frequently mix travel types (pedestrian, car, truck, bicycle). For example, we might set this penalty for matrix for pedestrians but not for cars in one VRP solution, and, again, it will depend on the specific case.

As @nilsnolde mentioned earlier, examples such as garbage collection or postman deliveries are typical. Postal workers, for instance, must adhere to the law, allowing street crossing only at crossroads (if there is any). They face fines for non-compliance, and in unfortunate incidents, they may encounter insurance-related issues.

Granted, commercial companies may not be as meticulous, and their drivers often adopt practices as mentioned by @kevinkreiser 
 stopping on one side and distributing deliveries on foot. However, this deviates from the planned solution and resembles a form of micro-crossdocking. This approach heavily relies on the nature of the goods being distributed; it's impractical for heavy packages or multiple bags of groceries. Therefore, you typically know, before computing travel times, whether you want a penalty for street crossing or not.

One consideration is the presence of pedestrian crossings (uncertain if Valhalla currently reads this data and its quality in OSM). If available, this information could serve as a valuable starting point to make the penalty application more sophisticated. Nonetheless, even with a crossing, there are cases when you might prefer staying on the same side and applying a penalty, even if there are legal pedestrian crossings.

A ZigZag street consumes more time, and that is the primary and original reason why postal workers avoid it. When planning for postal workers, it makes sense to have the option to impose a penalty for this type of crossing. When planning something else by default this penalty should be turned off.

I also don't think that incorporating street names into an algorithm for computing penalties would be beneficial. Right now I can't imagine a scenario where this information would be helpful, as the same or another street can be used to determine whether a penalty should or should not be applied.

2)
I cannot comment on TSP optimization in Valhalla, as we don't use it at all. In my opinion, this is more of a routing feature than an optimization feature. This distinction should be reflected in the (travel time value), which is derived from route/matrix results. I dont know how this works but probably in optimization request Valhalla computes matrix, and then some TSP algo applied. So this should be in matrix results and TSP algo naturaly use it, so zig zag happens only where penalty is less then traveltime to next building on same street side.

Summary:
IMO Simply having the ability to apply or not apply the penalty in request settings would be a significant improvement.

**Comment by dnesbitt61** [comment]
Could the optimizer utilize side of street for each location (not sure if this is returned for each source/target location within a matrix response)?

**Comment by nilsnolde** [comment]
good suggestion! side of street for sources & targets are not returned yet but the locations are already returned, adding the street side should be less work (and less request parameters, always good!). WDYT @radekvermirovsky ?
